### ISSUE TITLE
[Feature] Server side Infinite scrolling List Widget

### ISSUE THREAD

**Issue by jsartisan** [appsmithorg/appsmith#6293]
# Summary: 
As a dev, I should be able to load a list data in an async manner with infinite scrolling pagination.

# Requirements:

| Title | Description | Comment |
| ----- | ------ | ------- |
| Pagination type | As a dev, I should be able to choose if the list should be paginated with number controls or load with infinite scrolling | This is still under debate, It might not make sense to add another widget for this and also the list component commonly operates with infinite scrolling | 
| Items per page | How many items to render per api call. The size of the view port could automatically decide this | - | 
| Loading threshold | Defines the number of items pending in the viewport from the un-scrolled area to identify the time to call the api again. | - |

<img src="https://front.com/assets/img/favicons/favicon-32x32.png" height="16" width="16" alt="Front logo" /> [Front conversations](https://app.frontapp.com/open/top_5aj64)

**Comment by somangshu** [comment]
@Nikhil-Nandagopal @areyabhishek we think this should be a separate widget to avoid complications and maintenance issue. Do you think it makes sense from the UX standpoint?

**Comment by Nikhil-Nandagopal** [comment]
@somangshu I don't think that makes sense because this widget will be the most widely used form of the list widget. I also don't see why implementation should differ between these 2. In fact I am ok killing this way of pagination in favor of only scroll method because this method of pagination was a hack we introduced but list widgets generally don't work like this

**Comment by somangshu** [comment]
@riodeuno @jsartisan can you explain why we choose to create a separate widget for this technically?

Also I think I would agree to this being the default and only pagination for the list widget! Does this solve out problem better? Although this will required some additional feedback / poll.

**Comment by jsartisan** [comment]
We will go with implementing the virtualized list in the same widget. Thinking to toggle the virtualization with a toggle control.

**Comment by dilippitchika** [comment]
@somangshu is there another issue for the list widget to scroll instead of pagination for all data on the client side? This small fix would actually improve the UX for a lot of list implementations. 

Infinite scrolling can be done after we implement it for the table widget.

**Comment by somangshu** [comment]
@dilippitchika I did not catch what small fix you are talking about! 
I am also not sure if actually the table widget implementation for infinite scrolling will fix it for the list widget as well. 

I think once we iron out the initial big problems, we will be able to get to these

**Comment by dilippitchika** [comment]
Agreed I am just using this to kick off a discussion, what i was referring to was list ONLY having pagination today and no scroll. If i don't want pagination, there's no way out of it. 

Most list implementations have scroll and not pagination. However, we don't offer that choice to the user since we auto-paginate. We should allow a way to switch off pagination, because all items will simply show up in a scrollbar. 

Infinite scrolling design i believe will be similar to what we do in the table widget, not from an implementation POV. This is what i was referring to.

**Comment by divyadeepsawhney** [comment]
I agree with @dilippitchika. Pagination should be a choice, not a forced option. I just got stuck because of this while creating a comment feed. How can I increase the number of records per page on a list? @areyabhishek @Nikhil-Nandagopal

**Comment by somangshu** [comment]
@divyadeepsawhney thanks for the opinion :) 
As Dilip pointed out the pagination is automatic today, So to have more records per page you just need to increase the height of the list widget. The list widget tries to accommodate as many items as possible on the viewport based on its height, Post that it will paginate the rest of the items. Hope this answers your questions!

**Comment by lublak** [comment]
My thought is not only about asynchronous reloading. But also with data that can no longer be displayed in the list or table. With the list:  If there are more data than are displayed we can automatically display pagination. There is no possibility here to make a list scrollable which I would prefer.
![image](https://user-images.githubusercontent.com/44057030/186095894-9a607c19-025c-4e2a-9df0-9f3ca9ad4da1.png)
The same applies to the tables. Here I would also like to see a scroll functionality. Although you can deactivate the pagination here. But that doesn't necessarily make the table scrollable. Whereby here in general would have to be looked over again. Here is an example for 10 records. Here should be just when pagination is active no longer scrollable or?
![zfC5HnB1Fu](https://user-images.githubusercontent.com/44057030/186097456-c02b7317-0560-4a2a-a687-460f2ec6113c.gif)

**Comment by dilippitchika** [comment]
@lublak we are tackling it for the table widget here - https://github.com/appsmithorg/appsmith/issues/13180

Regarding the list widget yes we plan to support scrolling as well here

**Comment by Nikhil-Nandagopal** [comment]
@lublak by the way if you're not using server side pagination, you can still achieve this in the table by turning the server side pagination option on. It will make the table scrollable but there will be performance issues if you're loading a very large data set

**Comment by dilippitchika** [comment]
@SatishGandham @ashit-rath internal doc for PRD, we will pick up design this week. https://www.notion.so/appsmith/List-widget-infinite-scrolling-403e612eb40a4db6a1a2c1b64037e7ca

**Comment by shastryblr** [comment]
[Infinite scrolling designs ]

[figma](https://www.figma.com/file/aky4eMolJPQSUIBe9VzwW3/List-widget-Infinite-scrolling?node-id=200%3A7739)

loom & reasoning wip

**Comment by ogbofjnr** [comment]
Is there approximate date when this feature might be done?

**Comment by dilippitchika** [comment]
@ogbofjnr we are trying to pick this up early next year to deliver.

**Comment by honoru-washswat** [comment]
how did this end?

**Comment by m1ker1n** [comment]
I would like to have scrollable List Widget too

**Comment by laioncenet** [comment]
Hello Devs! I need help. How to create a vertical scrolling similar to a chat box for a live support system.

**Comment by wawa79** [comment]
Hello! Scrollable List Widget would be a nice feature. Pagination is just unusable when loading many records.

**Comment by Nikhil-Nandagopal** [comment]
@laioncenet vertical scroll is not yet supported.

@wawa79 you can paginate lots of records following this guide https://docs.appsmith.com/build-apps/how-to-guides/Setup-Server-side-Pagination-on-List

**Comment by laioncenet** [comment]
When will the inverted vertical scrolling feature for the table widget be launched?

**Comment by Nikhil-Nandagopal** [comment]
@laioncenet I'm not clear on what you mean by inverted vertical scrolling feature.

**Comment by laioncenet** [comment]
Okay, I'll give you an example: in the Retool tool, you have the option to create applications, and there is a template called AI Chat. Editing this model, the agent finds a chat message project, thus, when several messages are sent, a vertical scroll bar is created and it always descends automatically as messages are sent or the AI agent sends them to us.

So I think that in appsmith, would there be an option for the scroll bar instead of going up and down automatically? Do you have an estimate of whether this feature will be released?

I managed to create a chat app integrated with the WhatsApp Business API. However, the scroll bar keeps going up, so all new messages keep coming up.

The correct thing is for the messages to keep coming down.

What is the solution?

**Comment by jonathanroze** [comment]
Any news about this feature ? I would be a very nice feature... Waiting for!

**Comment by albseb511** [comment]
you can expose an event that reaches the bottom of the list element? 
onBottomReached, and maybe allow an offset from bottom. 
This can be achieved with Observation API.

**Comment by neilellis** [comment]
I agree with the previous comments about pagination, it's really clunky when you need to scan through a lot of items in a list. Infinite scroll is much more natural.

**Comment by parlt** [comment]
are there some updates? I need this feature too

**Comment by TC-Flowplan** [comment]
2026 and still no update?
